### PR TITLE
[FLINK-34256][doc] Add a documentation section for minibatch join

### PR DESCRIPTION
**Pull request by xishuaidelin**
## What is the purpose of the change

This pull request add doc for mini-batch join.

## Brief change log

Add doc for mini-batch join in Performance Tuning which belongs to table.

## Verifying this change

N/A

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

### CODE REVIEW

File: docs/content/docs/dev/table/tuning.md [hunk 1]
```diff
@@ -266,5 +266,23 @@ GROUP BY day
 Flink SQL optimizer can recognize the different filter arguments on the same distinct key. For example, in the above example, all the three COUNT DISTINCT are on `user_id` column.
 Then Flink can use just one shared state instance instead of three state instances to reduce state access and state size. In some workloads, this can get significant performance improvements.
 
+## MiniBatch Join
```
Comment (lsyldliu) [inline docs/content/docs/dev/table/tuning.md:269]: Please also update this sentence, which should contain the regular join: https://github.com/apache/flink/pull/24240/files#diff-8d75f38adf6c9f25a579a8fd856b1946386d4aae2ecdc8bbdb2c2d548ac2a590R31

```In this page, we will introduce some useful optimization options and the internals of streaming aggregation, regular join which will bring great improvement in some cases.```

**Review by lsyldliu — COMMENTED** [review]
@xishuaidelin Thanks for contribution, I left some comments.

File: docs/content.zh/docs/dev/table/tuning.md [hunk 2]
```diff
@@ -259,5 +259,23 @@ GROUP BY day
 
 Flink SQL 优化器可以识别相同的 distinct key 上的不同过滤器参数。例如，在上面的示例中，三个 COUNT DISTINCT 都在 `user_id` 一列上。Flink 可以只使用一个共享状态实例，而不是三个状态实例，以减少状态访问和状态大小。在某些工作负载下，可以获得显著的性能提升。
 
+## MiniBatch Join
+默认情况下，普通join算子是逐条处理输入的记录，即：（1）从状态中根据joinKey查询记录，（2）将当前记录写到对应状态，（3）处理输入和从状态中查找出的记录。这种处理模式可能会增加 StateBackend 开销（尤其是对于 RocksDB StateBackend ）。
```
[the remark this answers is not in the document]
Comment (xishuaidelin) [reply docs/content.zh/docs/dev/table/tuning.md:263]: done

File: docs/content/docs/dev/table/tuning.md [hunk 3]
```diff
@@ -266,5 +266,23 @@ GROUP BY day
...
+When the bundle of inputs is triggered to be processed, the inputs inside of the bundle are records that could not be folded further. Another optimization for update records is applied for the bundle. When encountering the pair of -U and +U records, they would be recognized and redundant records in their output would be suppressed. The graph below explains the principle here.
+
+{{< img src="/fig/table-streaming/suppress.jpg" width="70%" height="70%" >}}
+
+Besides, the order in which the left and right stream bundles are processed can also help reduce redundant records, but this only applies to cases where there is an outer join. The following graph clarifies the principle:
+
+{{< img src="/fig/table-streaming/order.jpg" width="70%" height="70%" >}}
+
+MiniBatch optimization is disabled by default for regular join. In order to enable this optimization, you should set options `table.exec.mini-batch.enabled`, `table.exec.mini-batch.allow-latency` and `table.exec.mini-batch.size`. Please see [configuration]({{< ref "docs/dev/table/config" >}}#execution-options) page for more details.
+The examples could refer to the part of MiniBatch Aggregation.
```
[the remark this answers is not in the document]
Comment (xishuaidelin) [reply docs/content/docs/dev/table/tuning.md:286]: Good idea. Done

File: docs/content/docs/dev/table/tuning.md [hunk 4]
```diff
@@ -266,5 +266,23 @@ GROUP BY day
 Flink SQL optimizer can recognize the different filter arguments on the same distinct key. For example, in the above example, all the three COUNT DISTINCT are on `user_id` column.
 Then Flink can use just one shared state instance instead of three state instances to reduce state access and state size. In some workloads, this can get significant performance improvements.
 
+## MiniBatch Join
+
+By default, regular join operator processes input records one by one, i.e., (1) look up records from state according to joinKey, (2) write or retract input in state, (3) process the input and joined records. This processing pattern may increase the overhead of StateBackend (especially for RocksDB StateBackend).
+
+The core idea of mini-batch join is to cache a bundle of inputs in a buffer inside of the mini-batch join operator. Reduce data in the cache, and then when the cache is triggered for processing, perform specific optimizations based on certain scenarios. Some of input records would be folded according to specified rule illustrated below:
```
[the remark this answers is not in the document]
Comment (xishuaidelin) [reply docs/content/docs/dev/table/tuning.md:273]: A new graph to clarify the principle is introduced.

File: docs/content/docs/dev/table/tuning.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (xishuaidelin) [reply docs/content/docs/dev/table/tuning.md:269]: Done

**Review by lsyldliu — APPROVED** [review]
LGTM

**Pull request closed without merging**